### ISSUE TITLE
rpath errors

### ISSUE THREAD

**Issue by vtjnash** [JuliaLang/julia#17602]
It seems like the recent rpath change may need to be reverted as it is causing frequent travis failures (usually with openspecfun, but also fftw):
https://travis-ci.org/JuliaLang/julia/jobs/146885109
https://travis-ci.org/JuliaLang/julia/jobs/146808597
https://travis-ci.org/JuliaLang/julia/jobs/146806235
https://travis-ci.org/JuliaLang/julia/jobs/146773370
https://travis-ci.org/JuliaLang/julia/jobs/146412731
https://travis-ci.org/JuliaLang/julia/jobs/145955397

**Comment by Keno** [comment]
Sounds like those libraries need to have their rpaths fixed.

**Comment by yuyichao** [comment]
Might be related. `optirun julia` also cause libraries in `/usr/lib/julia` fails to load.

**Comment by ViralBShah** [comment]
Do we need these rpath changes for 0.5? I suspect I read that was necessary but can't recollect. If not, perhaps best to revert for now and fix after branching.

**Comment by Keno** [comment]
We can add back the rpath to executable, but we shouldn't revert the entirety of the rpath changes.

**Comment by StefanKarpinski** [comment]
Libraries with failures above: libopenspecfun, libfftw3f, libopenlibm. So we should be in a position to fix these libraries' rpaths without having to argue to much with upstream (since we are upstream).

**Comment by ViralBShah** [comment]
These rpath fixes have to be in the julia build - not in the upstream (even though it is us).

**Comment by ViralBShah** [comment]
If I can  reproduce the issues, I might be able to help fix.

**Comment by Keno** [comment]
Ok, fftw3_threads seems to be accidentally setting RUNPATH instead of RPATH, which doesn't make much sense AFAIK, since it's not an executable. May need a patch to FFTW.

openspecfun needs an appropriate addition to it's link line, which is the following patch and the referenced PR to openspecfun.
```
diff --git a/Make.inc b/Make.inc
index 2264fea..91a6e31 100644
--- a/Make.inc
+++ b/Make.inc
@@ -845,10 +845,12 @@ ifeq ($(OS), WINNT)
 else ifeq ($(OS), Darwin)
   RPATH := -Wl,-rpath,'@executable_path/$(build_libdir_rel)'
   RPATH_ORIGIN := -Wl,-rpath,'@loader_path/'
+  RPATH_ESCAPED_ORIGIN := $(RPATH_ORIGIN)
   RPATH_LIB := -Wl,-rpath,'@loader_path/' -Wl,-rpath,'@loader_path/julia/'
 else
   RPATH := -Wl,-rpath,'$$ORIGIN/$(build_libdir_rel)' -Wl,-rpath-link,$(build_shlibdir) -Wl,-z,origin
   RPATH_ORIGIN := -Wl,-rpath,'$$ORIGIN' -Wl,-z,origin
+  RPATH_ESCAPED_ORIGIN := -Wl,-rpath,'\$$\$$ORIGIN' -Wl,-z,origin
   RPATH_LIB := -Wl,-rpath,'$$ORIGIN' -Wl,-rpath,'$$ORIGIN/julia' -Wl,-z,origin
 endif

diff --git a/deps/openspecfun.mk b/deps/openspecfun.mk
index 0df44fd..781e47e 100644
--- a/deps/openspecfun.mk
+++ b/deps/openspecfun.mk
@@ -11,7 +11,7 @@ endif

 OPENSPECFUN_OBJ_TARGET := $(build_shlibdir)/libopenspecfun.$(SHLIB_EXT)
 OPENSPECFUN_OBJ_SOURCE := $(BUILDDIR)/$(OPENSPECFUN_SRC_DIR)/libopenspecfun.$(SHLIB_EXT)
-OPENSPECFUN_FLAGS := ARCH="$(ARCH)" CC="$(CC)" FC="$(FC)" AR="$(AR)" OS="$(OS)" USECLANG=$(USECLANG) USEGCC=$(USEGCC) FFLAGS="$(JFFLAGS)" CFLAGS="$(CFLAGS) $(OPENSPECFUN_CFLAGS)"
+OPENSPECFUN_FLAGS := ARCH="$(ARCH)" CC="$(CC)" FC="$(FC)" AR="$(AR)" OS="$(OS)" USECLANG=$(USECLANG) USEGCC=$(USEGCC) FFLAGS="$(JFFLAGS)" CFLAGS="$(CFLAGS) $(OPENSPECFUN_CFLAGS)" LDFLAGS="$(RPATH_ESCAPE

 ifeq ($(USE_SYSTEM_LIBM),0)
        OPENSPECFUN_FLAGS += USE_OPENLIBM=1
```

**Comment by Keno** [comment]
I'll put up a quick fix to add the rpath to both libraries after the fact.

**Comment by Keno** [comment]
Hmm, after doing some more research, the RUNPATH setting in fftw should actually be fine (contrary to what the dlopen man page says). I was also unable to reproduce the error seen on CI wrt fftw.

**Comment by tkelman** [comment]
Do you have a system copy of fftw installed? I think it's likely that Travis VM's do, as a dependency of something else that they have installed by default.

**Comment by ViralBShah** [comment]
We can always uninstall the system one on travis before doing julia stuff.

**Comment by tkelman** [comment]
I'm actually not sure how to get Travis to uninstall things in the docker container, since we can't do sudo and all apt-get operations have to go through the `apt: packages` whitelist thing. They may allow caching now in the sudo-true workers so we could look into switching - but not right now.

**Comment by tkelman** [comment]
PR incoming

**Comment by Keno** [comment]
@tkelman Yes, you're correct. I had a global fftw installed. Still, I can't reproduce the fftw failure. I guess we'll see if it happens again on CI, but it definitely doesn't seem as straightforward as "missing rpath" since it does have RUNPATH set.

**Comment by tkelman** [comment]
maybe runpath is buggy back with ubuntu 12.04 toolchains? we're using the ubuntu-toolchain-r-test ppa's gcc-5 package, but dunno if that rebuilds binutils or other core things.

**Comment by vtjnash** [comment]
It looks like fftw is missing the "I have the string $ORIGIN in my -rpath" flag (the `-Wl,-z,origin` option to ld)?

```
~/julia/usr/lib$ readelf -d libfftw3_threads.so  | grep ORIGIN
 0x0000001d (RUNPATH)                    Library runpath: [$ORIGIN]
```

**Comment by vtjnash** [comment]
also, not sure that libmpfr is built correctly, since it's missing rpath link to ensure we load the right libgmp:
```
jameson@jamesonnash:~/julia/usr/lib$ readelf -d libmpfr.so

Dynamic section at offset 0x5ae60 contains 27 entries:
  Tag        Type                         Name/Value
 0x00000001 (NEEDED)                     Shared library: [libgmp.so.10]
 0x00000001 (NEEDED)                     Shared library: [libc.so.6]
 0x00000001 (NEEDED)                     Shared library: [ld-linux.so.2]
 0x0000000e (SONAME)                     Library soname: [libmpfr.so.4]
```

**Comment by vtjnash** [comment]
Apparently the documentation for RPATH is wrong / incomplete? http://blog.qt.io/blog/2011/10/28/rpath-and-runpath/

**Comment by Keno** [comment]
Yes, it is. I do believe having a RUNPATH in fftw is sufficient though (from what I can tell the difference between RPATH and RUNPATH is that RPATH applies recursively). I've been contemplating updating the dlopen man page to correct it.

**Comment by vtjnash** [comment]
Right, but without the `-z origin` flag during linking, that may be specifying the literal directory "usr/lib/\$ORIGIN" instead of referencing the `origin` location? And since it's using RUNPATH, the fallback path to the ./julia RPATH is disabled (which is possibly what's making mpfr work, even though it might not be linked right either)

**Comment by vtjnash** [comment]
that flag is also missing from libarpack:
```
jameson@jamesonnash:~/julia/usr/lib$ readelf -d libarpack.so | grep ORIGIN
 0x0000001d (RUNPATH)                    Library runpath: [$ORIGIN]
```

expected result should contain:
```
 0x6ffffffb (FLAGS_1)                    Flags: ORIGIN
```

**Comment by Keno** [comment]
I believe we're setting RUNPATH using `patchelf`. What's the effect of not having `-z origin`?

**Comment by vtjnash** [comment]
AFAICT, it means the linker won't replace `$ORIGIN` with the origin value (it'll treat it as a literal instead)

(`libgit2`, `libmbed*`, and `libssl` also all appear to be similarly broken)

**Comment by vtjnash** [comment]
nvm, seems like `-z origin` should be a non-functional flag ("DF_ORIGIN") on linux:
http://www.sco.com/developers/gabi/2003-12-17/ch5.dynamic.html "Substitution Sequences"

**Comment by vtjnash** [comment]
Should we also verify / fix libarpack, libmpfr, libgit2 & dependencies to have a DT_RPATH entry before we call this fixed?

**Comment by tkelman** [comment]
yeah i tried to reopen it but then github auto-closed it a second time...

**Comment by tkelman** [comment]
@Keno there is still a linux issue here. Try installing a nightly in a container where there's no system-wide libgfortran, you get `could not load library "libopenblas64_" libgfortran.so.3: cannot open shared object file: No such file or directory`.

**Comment by Keno** [comment]
Do we bundle libgfortran?

**Comment by tkelman** [comment]
yes

**Comment by tkelman** [comment]
do we need to invoke patchelf in https://github.com/JuliaLang/julia/blob/master/contrib/fixup-libgfortran.sh ?

**Comment by tkelman** [comment]
Bump @Keno we need to get rid of that warning before 0.5.0 final.

**Comment by tkelman** [comment]
@Keno thanks for https://github.com/JuliaLang/julia/pull/17901, that does fix the gfortran warning when running rc2 binaries in a container that doesn't have a system-wide gfortran. This is no longer blocking but there may still be cleanup needed? Not sure what the current failure mode is, if any.

**Comment by vtjnash** [comment]
Adding this to the build milestone as it seems this means our binaries are unreliable? Libraries that seem to be missing `(RPATH)`:
 - [ ] libmpfr.so
 - [ ] libcrypto.so.1.0.0
 - [ ] libssl.so.1.0.0
 - [ ] libfftw3f.so / libfftw3.so
 - [ ] libdSFMT.so
 - [ ] libopenblas.so
 - [ ] libccalltest.so

**Comment by vtjnash** [comment]
x-ref https://github.com/JuliaLang/julia/issues/18106

**Comment by tkelman** [comment]
propose a fix now or it's not holding up 0.5.0.

**Comment by vtjnash** [comment]
IIUC from the travis failures, this would manifest as opening the wrong ssl library (causing mysterious failures in Pkg) and gmp (from mpfr, for BigFloat support). The fix is to run patchelf on the libraries or pass `-rpath` as a flag as Keno did for some of the other dependencies.

**Comment by tkelman** [comment]
As a PR if you know what's wrong, unless the issue is reliably relroducible.

**Comment by andreasnoack** [comment]
When building 0.5 on a cluster, I get that
```
libgit2/libgit2.jl
LoadError("sysimg.jl",326,LoadError("libgit2/libgit2.jl",13,LoadError("libgit2/consts.jl",282,ErrorException("error compiling version: could not load library \"libgit2\"\nlibssl.so.1.0.0: cannot open shared object file: No such file or directory"))))
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
```
and
```
[noack@node008 julia]$ ldd ldd /scratch/users/noack/julia/usr/lib/libgit2.so
	linux-vdso.so.1 =>  (0x00002aaaaaacb000)
	libcurl.so.4 => /scratch/users/noack/julia/usr/lib/libcurl.so.4 (0x00002aaaaafcb000)
	libz.so.1 => /lib64/libz.so.1 (0x00002aaaab244000)
	libssl.so.1.0.0 => not found
	libcrypto.so.1.0.0 => not found
	libssh2.so.1 => /scratch/users/noack/julia/usr/lib/libssh2.so.1 (0x00002aaaab45b000)
	librt.so.1 => /lib64/librt.so.1 (0x00002aaaab68d000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00002aaaab895000)
	libc.so.6 => /lib64/libc.so.6 (0x00002aaaabab3000)
	libmbedtls.so.10 => /scratch/users/noack/julia/usr/lib/libmbedtls.so.10 (0x00002aaaabe47000)
	libmbedx509.so.0 => /scratch/users/noack/julia/usr/lib/libmbedx509.so.0 (0x00002aaaac06d000)
	libmbedcrypto.so.0 => /scratch/users/noack/julia/usr/lib/libmbedcrypto.so.0 (0x00002aaaac281000)
	/lib64/ld-linux-x86-64.so.2 (0x00002aaaaaaab000)
```
Is that related to this issue?

**Comment by JeffBezanson** [comment]
It seems this has been worked around enough that it's no longer high priority?

**Comment by ViralBShah** [comment]
Should we have a final pass in the `binary-dist` target that rewrites `rpath` (or does whatever the right thing is) for all libraries we build and ship?

**Comment by StefanKarpinski** [comment]
Is this really a 1.0 blocker? How would fixing this break user code?

**Comment by JeffBezanson** [comment]
Would be great to fix in 1.0, but not a blocker.

**Comment by ViralBShah** [comment]
I believe we reliably set RPATH on all the libraries now. Please reopen if necessary.
